### DISCUSSION TITLE
IEA15MW with ROSCO wind turbine, ramped wind

### DISCUSSION THREAD

**Discussion by borjanavascues** [NatLabRockies/ROSCO#274] [General]
Good morning, 
My name is Borja and I am currently in the process of validating and benchnmarking the IEA15MW wind turbines models for my company, and to do this, I am checking the fixed furbine configuration. 

I am using the ROSCO V2.6.0 controller with the .IN file corresponding to the fixed monopile configuration. I started running some basic steady wind cases for a set of wind speeds, and I am encountering some issues in the Reg.II of the controller (corresponding to below-rated winds). If I use a **ramped wind** to avoid any initial instability, the **pitch of the blades don't match the results** published on the IEA15MW wind turbine repport, as the **pitch is set at around 3.4 degs. and doesn't change** once steady codnitions are encountered. On the other hand, if I run the **simulations with no ramping** time, the **pitch reaches its correct value** of 0 degs. for this region of the controller. This is the case for all the wuind velocities withing Reg.II of the controller. 

Is this normal? Why does this happen and how can I avoid it? Shouldn't the turbine properties be equal once the steady state is reached?

Please, find in the link below two screenshots corresponding to my simulations with and without ramping time for a wind speed of **8.22 m/s** in case you need them. 

[https://drive.saitec.es/1681/28ee86bd287c8ba](url)

Thank you in advance.

**Comment by dzalkind** [comment]
Hi, please post figures directly here.  I have seen this kind of behavior very occasionally due to the nonlinearity of the controller and wind speed estimator.  Updating to the latest version (v2.8.0) is also advisable. 

The minimum pitch at low wind speed is based on the wind speed estimate and the minimum pitch limit.  You can plot these signals (`WE_Vw_F` and `PC_MinPit`) from the RO.dbg2 files outputted when LoggingLevel = 2 in the DISCON.

**Reply by borjanavascues** [reply]
Hi, sorry, I dind't know figures could be directly attached here. Please, find attached the OrcaFlex figures mentioned above. 
Also, you can find the **plots of the controll variables** regarding the wind speed estimate and the minimum pitch limit. It seems like the **velocity estimate for the Ramped** simulation is **kept at 0** throughout the entire simulation, which may explain the different behavior; the minimum pitch value also differs. 

![IEA15MW-8_22-ramp](https://github.com/NREL/ROSCO/assets/148865919/9d1f1e36-a7a3-4a4d-ab56-fb4a326748d2)
![IEA15MW-8_22-NORAMP](https://github.com/NREL/ROSCO/assets/148865919/b508899d-0406-42fb-8704-4855686d007f)
![RAMP_NORAMP_FLAGS](https://github.com/NREL/ROSCO/assets/148865919/9d4d14b6-70ce-44ff-bc5b-dea161305949)

Is this normal? Why is the controller estimating a constant zero velocity? How can this be fixed?

Thanks for the quick response!

**Reply by dzalkind** [reply]
This may be an old bug in the wind speed estimator.  We previously saw it in long LES simulations with very low wind speeds.  A slow ramp-up might result in similar behavior.  Please update your ROSCO dynamic library to the latest version and try again.  Here is a tool that helps update DISCONs: https://github.com/NREL/ROSCO/blob/main/Examples/19_update_discon_version.py

**Reply by borjanavascues** [reply]
Good morning, and again, thank you very much for your fast response!

I had already run the simulations for both constant and ramped winds (10m/s this time) with the new **ROSCO V2.8.0** availble in the releases section of this repository, and with the **most updated .IN files** for the IEA15MW monopile configuration availble in its repository. **The problem seems to persists** if this updated version is used (as it can be observed in the graphs). 

![RAMP_NORAMP_FLAGS_V280](https://github.com/NREL/ROSCO/assets/148865919/6ff14864-428a-4842-8efb-732b6147e720)

Thinking that the configuartion files may be outdated, I run the python update script changing the inputs and paths to make it work with the IEA15MW turbine, and the updated version of the **configuration file doesn't change** with respect to the one availble on the IEA repository, only on the path of the performance tables file. 

If you need my controller, .IN, debug, or OrcaFlex files, I am happy to providing them so it's easier for you to check it.

**Reply by dzalkind** [reply]
Hi, I'm sorry, I looked into it and the bug was fixed in the `develop` branch, not v2.8.0.  If you compile the dynamic library in the develop branch and use those inputs, your setup should work.  Alternatively, if you start the ramp at 1 m/s (or something greater than 0), that should also work.

**Reply by borjanavascues** [reply]
Thank you very much. I've tried with an non-zero inital value and the controlers seems to work properly. 

How can we be sure that this bug doesn't affect the overall behaviour of the controller (e.g. with a fluctuating wind which has speeds of 0 m/s)? Does the bug only affects the initial behavour of the turbine?

Again, thank you very much for your help.

**Reply by dzalkind** [reply]
The issue with the actual wind speed affecting the wind speed estimator is isolated to the initial timestep.  In the `develop` branch, the minimum initial wind speed is limited to a minimum of 3 m/s.  There are saturation limits on the inputs of the WSE which should, in theory, prevent negative wind speed estimates because the wind speed estimate is only valid in operational wind speeds.  Ideally, with these saturation limits, the WSE will never return a non-operational wind speed.
